JBPM-5538 & JBPM-5449: Popups for unsaved data and no changes since last save

Christopher-Chianelli opened the pull request:
Depends on #910, which depends on kiegroup/droolsjbpm-build-bootstrap#449. The unsaved changes message will not appear if you make any changes now. Additionally, if you try saving without making any changes, a popup will appear informing you that there are no changes since you last saved. Some components hashCode functions need to be updated so that it changes when the component changes.

romartin commented:
ok to test

romartin commented:
Hey @Christopher-Chianelli , first of all thanks for this. 
I just realized now that this PR exists hehe, feel free for next times, when creating PRs for Stunner, to add a reference to my github user in the comment, this way I will receive the notification and will start as soon as possible the review :) Up to you!
Thanks!

hasys commented:
Jenkins execute full downstream build

Christopher-Chianelli commented:
@romartin I added equals and hashCode functions for the remaining beans, and wrote tests for saving, HashUtils, and equals/hashCode.

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/graph/impl/GraphImpl.java:
```diff
@@ -59,4 +69,73 @@ public Node getNode(final String uuid) {
+                            }
+
+                            @Override
+                            public void endEdgeTraversal(final Edge edge) {
```
can be removed - no need  to override this method, as already exist a default method on the interface

Christopher-Chianelli replied:
@romartin done! The two methods that just called the default were there just in case I need to modify them; since I don't they were removed. The multiple if statements where there since before for some reason my hashCode implementation for the nodes weren't being called. I have squashed all the commits. Note that this branch is now based at HEAD~1, since the current master branch is not compiling due to missing methods.

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/graph/impl/GraphImpl.java:
```diff
@@ -59,4 +69,73 @@ public Node getNode(final String uuid) {
+                            int[] myHashArr = hashArr;
+                            @Override
+                            public void
+                            startGraphTraversal(final org.kie.workbench.common.stunner.core.graph.Graph graph) {
```
can be removed - no need  to override this method, as already exist a default method on the interface

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/graph/impl/GraphImpl.java:
```diff
@@ -59,4 +69,73 @@ public Node getNode(final String uuid) {
+                                super.startEdgeTraversal(edge);
+                                final Object content = edge.getContent();
+                                myHashArr[0] = HashUtil.combineHashCodes(myHashArr[0], content.hashCode());
+                                if (content instanceof Child) {
```
I think that the different hash values combined inside these `if` sentences are not necessary, so in this method only combining the hash for the `edge.getContent()` should be enough. I think it's true because both sources and target nodes for each edge instance, are already being processed in the other callback method `startNodeTraversal`, so probably you're duplicating the hashes twice this way. 
Does it makes sense? Can you try it please?

romartin commented:
CC @manstis @hasys in case you want to review too :)

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/util/HashUtil.java:
```diff
@@ -0,0 +1,19 @@
+package org.kie.workbench.common.stunner.core.util;
```
License

manstis commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/test/java/org/kie/workbench/common/stunner/bpmn/definition/HashCodeAndEqualityTest.java:
```diff
@@ -0,0 +1,334 @@
+/*
+ * Copyright 2016 Red Hat, Inc. and/or its affiliates.
```
2017

jomarko commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/diagram/DiagramImpl.java:
```diff
@@ -28,4 +29,24 @@ public DiagramImpl(final @MapsTo("name") String name,
     }
+
+    @Override
+    public int hashCode()
```
This `hashCode()` and `equals()` below are kind of suspicious for me. Are we sure that `getGraph()` return always non null value? Same question about `getMetadata()` and `getName()`.

Christopher-Chianelli replied:
Thanks for finding that! Before, HashUtil.combineHashCodes took an array of Objects and called hashCode on each of them (after doing a null check). However, it always called Object's hashCode implementation, so I changed it to take an array of ints.

hasys commented:
Jenkins execute full downstream build

hasys commented:
Jenkins execute full downstream build

Christopher-Chianelli commented:
Reformatted code, added/updated copyright and added more null checks.

Christopher-Chianelli commented:
Resolved; we inserted packages at the same line. (I might have forgotten to alphabetize the order though...)

Christopher-Chianelli commented:
(now alphabetized)

hasys commented:
Jenkins execute full downstream build

romartin commented:
ok to test

hasys commented:
Jenkins execute full downstream build

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-api/src/main/java/org/kie/workbench/common/stunner/project/diagram/impl/ProjectDiagramImpl.java:
```diff
@@ -35,24 +36,27 @@ public ProjectDiagramImpl(final @MapsTo("name") String name,
+            return false;
+        }
+    }
+
```
Not tested methods.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java:
```diff
@@ -610,4 +622,8 @@ private void log(final Level level,
+
+    private boolean hasUnsavedChanges() {
+        return getCurrentDiagramHash() != originalHash;
+    }
```
No tests for changes in the file.

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java:
```diff
@@ -256,7 +262,12 @@ public void onError(final Collection<DiagramElementViolation<RuleViolation>> vio
+                                continueSaveOnceValid.execute();
+                            } else {
+                                //TODO: Localize message
+                                errorPopupPresenter.showMessage("You have no unsaved changes.");
```
I see duplicated not localized literal. @romartin how do we handle it for now?

romartin replied:
yes sure, please inject the `TranslationService` so you can use errai i18n here. Thanks!!  :)

romartin commented:
Hey @Christopher-Chianelli , please can you verify that latest included BPMN bean, the `EmbeddedSubprocess`, does also contains the right `equals/hasCode` methods? This new bean has been added a few days ago, before this commit, so probably it needs to override that methods too. 
Thanks!

Christopher-Chianelli commented:
I cannot really check anything that been added in the last commit, @romartin; the last commit fails to build on master due to missing classes (BlobImpl, I think, from uberfire). I can add the hashCode/equals, but I won't be able to test or run it on the webclient (compilation fails during gwt compile, not mvn build).

romartin commented:
Hey @Christopher-Chianelli , mmm the build works for me, but yeahh not sure why Errai snapshots seem not to be building on nexus for last few days... anyway building it locally makes it work, so please just build latest [Errai](https://github.com/errai/errai) and [Uberfire](https://github.com/AppFormer/uberfire) locally and you will be able to build & run the app. Lemme know if any issues! Thanks!

Christopher-Chianelli commented:
That fixed it! Thanks @romartin!

Christopher-Chianelli commented:
Added bunch of tests, made moving nodes count as changes and did localization (I think, notify me if I made a mistake). Also added hashCode and equals for `EmbeddedSubprocess`.

romartin commented:
Jenkins execute full downstream build

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java:
```diff
@@ -486,6 +493,17 @@ protected boolean _onMayClose() {
+            super.onSave();
+        }
+        else {
+            //TODO: Localize message
```
AFAIK this TODO can be removed

Christopher-Chianelli replied:
I only wrote the message in the English Common Constants properties file, so I don't know how it'll behave in other languages.

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java:
```diff
@@ -486,6 +493,17 @@ protected boolean _onMayClose() {
+        }
+        else {
+            //TODO: Localize message
+            Window.alert(CommonConstants.INSTANCE.NoChangesSinceLastSave());
```
Please better using the method `showError` (in this class), it uses an uberfire generic popup widget - we should avoid the use of  `Window.alert`

Christopher-Chianelli replied:
Do you want it to be an error or notification? @hasys requested for it to be a notification (since I couldn't find a notification class that is applicable to what data I have at the time of the popup, so I used `WIndow.alert` for a generic notification).

hasys replied:
Yes, sorry, I mean something like this: 
![image](https://user-images.githubusercontent.com/1477262/27485277-6bd39f22-582c-11e7-8c98-1d0b2debcc26.png)

Just notification, no Message Dialog with buttons.

hasys commented:
Jenkins execute full downstream build

hasys commented:
Jenkins execute full downstream build

Christopher-Chianelli commented:
Changed `Window.alert` to an info notification and removed the TODO @romartin @hasys.

hasys commented:
Jenkins execute full downstream build

hasys commented:
Jenkins execute full downstream build

Christopher-Chianelli commented:
Thanks for informing me @hasys! I'll start investigating the issue immediately.

Christopher-Chianelli commented:
Fixed issue @hasys, should I just push the fix to this branch?

hasys commented:
Hi @Christopher-Chianelli, for me two small PR are better than one huge. If it's possible I prefer merge this one (it is already big one) and send new PR. Thank you!
